### PR TITLE
Refactor and add org-level project support, plus tests

### PR DESCRIPTION
**Pull request by richkuz**
Hello @alexh97 , would you be willing to review this PR and consider merging it? There is no rush on this, it is just a small personal spacetime experiment. I am able to use my forked action without a problem, but I thought I'd offer these improvements back to the origin project.

## Motivation: Org-wide project support

I want to use GitHub actions to automatically add issues and PRs to _organization-wide_ projects. The current `project-assigner` action is limited to projects that live in the same repo as the associated issue or PR.

The crux of the problem is that when adding a label to a project, the `project-assigner` only queries for column IDs within a repo-wide project: https://github.com/elastic/github-actions/blob/master/project-assigner/index.js#L34

I added a new field to `issue-mappings` called `projectScope`. We can now find columns in projects at the organization, user, and repository scope: https://github.com/richkuz/github-actions/blob/master/project-assigner/project-assigner.js#L42

I was able to create a [GitHub action](https://github.com/elastic/app-search-team/blob/master/.github/workflows/project-assigner.yml) that assigns [app-search-team repo](https://github.com/elastic/app-search-team) issues to an [Elastic org-wide test project](https://github.com/orgs/elastic/projects/516).

## Extra: Refactoring and unit tests

I found I could not easily test my changes with confidence. So I refactored the existing code into a class based structure to make it easier to unit test the critical code paths. I performed [manual testing](https://github.com/richkuz/github-actions/blob/master/project-assigner/README.md#manual-testing) on my own local projects, too.

I also corrected some typos in the Readme and improved the documentation around setting a GitHub token with bigger permissions.

The changes should be backwards-compatible with previous versions, hence only a minor version bump is required (`2.1.0`).

To review this PR, it is probably best to start with the documentation changes first to get a feel for the change. Then read the `handleLabeled` function and the functions it calls. Next, read the `handleUnlabeled` function and the functions it calls. Lastly, read through the unit tests. Unfortunately, diffing the new source with the old source is not easy because of the refactoring.

Output from `npm test`:
```
  projectAssigner
    handleLabeled
      ✓ creates a project card when a specific label is added to an issue (34 ms)
      ✓ does not create a project card when a card already exists for the issue (1 ms)
      ✓ does not create a project card when the label doesn't match (1 ms)
      ✓ requires a projectNumber, columnName and projectScope (17 ms)
    handleUnlabeled
      ✓ removes a project card when a specific label is removed from an issue (1 ms)
      ✓ does not remove a project card when the issue's project number doesn't match (1 ms)
      ✓ does not remove a project card when a card does not exist for the issue (1 ms)
      ✓ does not remove a project card when the label doesn't match
      ✓ requires a projectNumber
    normalizeGithubContext
      ✓ returns a normalized context for an issue event (1 ms)
      ✓ returns a normalied context for a pull request event
    removeCard
      ✓ removes a card (1 ms)
      ✓ throws an error when attempting to remove a card with an invalid card ID (1 ms)
    findProjectCardsForPayloadItem
      ✓ finds project cards for an existing project issue (2 ms)
      ✓ finds project cards for an existing project pull request (1 ms)
      ✓ finds no project cards for an issue assigned to no projects
      ✓ throws an error when trying to find project cards for an issue that does not exist (1 ms)
    createCard
      ✓ creates a card (2 ms)
      ✓ throws an error when attempting to create a card with an invalid issue content ID
    findColumnIdForColumnName
      find columns in projects with different projectScope
        ✓ finds a column in a repo project (1 ms)
        ✓ finds a column in an organization project (1 ms)
        ✓ finds a column in a user project
      failure cases when finding columns
        ✓ fails to find a column when given a valid project number and missing column name (1 ms)
        ✓ fails to find a column when given an valid project number
    findProjectCardId
      ✓ finds a project card for an issue (1 ms)
      ✓ finds a project card for a pull request (1 ms)
      ✓ fails to find a project card for an issue in a different project (1 ms)
```

### CODE REVIEW

File: project-assigner/project-assigner.js [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            case 'org':
+                return `organization(login: "${context.owner}")`;
+            case 'user':
+                return `user(login: "${context.owner}")`;
+            case 'repo':
+                return `repository(owner: "${context.owner}", name: "${context.repo}")`;
+            default:
+                throw new Error(`Invalid projectScope ${projectScope}. Expected: org, user, or repo`);
+        }
+    }
```
Comment (goodroot) [inline project-assigner/project-assigner.js:141]: Oh heck yeah. 😆

**Comment by ph** [comment]
@richkuz 

I've been trying this PR on   https://github.com/elastic/fleet-server but I can't seem to get it to work, I get the following error when the action is run.

```
 Action failed with error: Error adding Issue to org project 478: column "Inbox" was not found
```

My configuration look like this 

https://github.com/elastic/fleet-server/blob/d6e8cd3640dc402588f79a8bab99caf327f9dc94/.github/workflows/project-assigner.yml#L18

Both the project and the column exists, I just added the `public_repo` permission on the token, if it was a permission issue I would receive something else.

**Review by alexh97 — APPROVED** [review]
So sorry it took me so long to get to this review!
This looks awesome and I'm glad you were able to figure out how to find the right scope for the fields in GH API. Their documentation wasn't the best when I wrote the initial version.

**Comment by richkuz** [comment]
Thanks @alexh97 !

@ph I took a quick glance at the action logs, your configuration looks correct. And I confirmed in the [GraphQL Explorer](https://docs.github.com/en/graphql/overview/explorer) that the query it's attempting to perform returns the Inbox column:
```
{
                organization(login: "elastic") {
                    project(number: 478) {
                        columns(first: 50) {
                            nodes {
                                name,
                                id
                            }
                        }
                    }
                }
            }
```
Returns:
```
{
  "data": {
    "organization": {
      "project": {
        "columns": {
          "nodes": [
            {
              "name": "Inbox",
              "id": "MDEzOlByb2plY3RDb2x1bW4xMjU1MTUwMg=="
            },
...
```

There must be a bug somewhere around https://github.com/richkuz/github-actions/blob/master/project-assigner/project-assigner.js#L94 . I will take a look tomorrow and add some more debug logging. I'll wait to resolve this issue before merging this PR.

**Comment by richkuz** [comment]
Hmm. I created a [unit test](https://github.com/richkuz/github-actions/commit/06eb68cd2d9df4916f60b2ba30b59b61eeade49b) with these exact mocked payloads and it worked.

@ph Can you please bump your version of project-assigner to 2.1.1 from 2.1.0? I added some debug logging that will help me diagnose it further on your actual instance.

**Comment by ph** [comment]
@richkuz 

doesn't seems to work for me, 
```
Run richkuz/github-actions/project-assigner@v2.1.1
Query for project cards:
{
                repository(owner: "elastic", name: "fleet-server") {
                    issue(number: 344) {
                        projectCards {
                            edges {
                                node {
                                    project {
                                        number
                                    },
                                    id
                                }
                            }
                        }
                    }
                }
            }
Response from query for project cards:
{
  "repository": {
    "issue": {
      "projectCards": {
        "edges": []
      }
    }
  }
}
Query for project columns:
 {
                organization(login: "elastic") {
                    project(number: 478) {
                        columns(first: 50) {
                            nodes {
                                name,
                                id
                            }
                        }
                    }
                }
            }
Response for project columns query:
 {
  "organization": {
    "project": null
  }
}

**Comment by richkuz** [comment]
@ph noted this error after changing the security key:
`GitHub Actions has encountered an internal error when running your job.`

This error might appear when the repository secret is defined with a newline character in it. When GitHub actions replaces `${{ secrets.ELASTIC_ORG_GITHUB_ACTIONS_TOKEN }}` in `project-assigner.yml` with the security token, if the token contains an extra newline or other invalid YML character, the resulting YML configuration could cause the action to fail in this manner.

I tested the specific project and repo with my own key and confirmed the action is able to retrieve the column data. It's likely the issue @ph is having is with the security token.

I'm going to merge this PR. We can troubleshoot this specific case as a separate issue further if needed.

**Comment by ph** [comment]
I was able to make it work this morning! Thanks @richkuz

**Comment by jportner** [comment]
@richkuz @alexh97 it looks like this was added in a v2.1.2 release in Rich's repo (https://github.com/richkuz/github-actions/releases), but this repo's latest release is v2.0.0.

Who's in charge of cutting new releases in this repo? Can someone make one so we can start using this in Kibana? (elastic/kibana#105307)

**Comment by alexh97** [comment]
@tylersmalley just tagged a v2.1.0 with @richkuz 's latest changes.

**Comment by tylersmalley** [comment]
I went ahead and created a v2.1.0 release, which also coincides with the [README](https://github.com/elastic/github-actions/tree/master/project-assigner#a-workflow-configuration-for-assigning-issues-to-projects) https://github.com/elastic/github-actions/releases/tag/v2.1.0

**Comment by jportner** [comment]
@richkuz (sorry for the double ping) it looks like you have an extra commit in your repo that you released as v2.1.2 (https://github.com/richkuz/github-actions/commit/03129580144c535949134ea3d581f629df7ec462)

Is that important, and is it stable? If so, perhaps it would be a good idea to merge it upstream here so we can cut a v2.1.2 release on the elastic repo?

**Comment by richkuz** [comment]
@jportner it's stable, but only mildly important. IIRC, when rapidly adding/removing a label to the same issue in a repo, the GH action could fail with an error. I'll open a PR to merge it upstream here.

**Comment by richkuz** [comment]
https://github.com/elastic/github-actions/pull/39

**Pull request merged**